While promises are often associated with asynchronous operations, they are simply placeholders for values. The value may already be known or, more commonly, the value may be the result of an asynchronous operation. Instead of subscribing to an event or passing a callback to a function, a function can return a promise, like this:

```js
// fetch() promises to complete at some point in the future
const promise = fetch("books.json");
```

The fetch() function is a common utility function in JavaScript runtimes that makes network requests. The call to fetch() doesn’t actually complete a network request immediately; that will happen later.
Instead, the function returns a promise object (stored in the promise variable in this example, but you can name it whatever you want) representing the asynchronous operation so you can work with it in the future. Exactly when you’ll be able to work with that result depends entirely on how the promise’s lifecycle plays out.

## The Promise Lifecycle

Each promise goes through a short lifecycle starting in the pending state, which indicates that promise hasn’t completed yet. **A pending promise** is considered **unsettled**
Once the **promise completes**, the promise is considered **settled** and enters one of two possible states:
1. Fulfilled: The promise has completed successfully.
2. Rejected: The promise didn’t complete successfully due to either an error or some other cause.

![The_Promise_Lifecycle](The_Promise_Lifecycle.png)

An internal **PromiseState** property is set to "pending", "fulfilled", or "rejected" to reflect the promise’s state. This property isn’t exposed on promise objects, so you can’t determine which state the promise is in programmatically

### Assigning Handlers with `then()`

The `then()` method is present on all promises and takes two arguments:
	- The first argument is a function to call when the promise is fulfilled, called the fulfillment handler
	- The second argument is a function to call when the promise is rejected, called the rejection handler
Both arguments to then() are optional, so you can listen for any combination of fulfillment and rejection.

```js
const promise = fetch("books.json");

// add a fulfillment and rejection handler
promise.then(response => {
// fulfillment
console.log(response.status);
}, reason => {
// rejection
console.error(reason.message);
});

//add another fulfillment handler
promise.then(response => {
// fulfillment
console.log(response.statusText);
});

//add another rejection handler
promise.then(null,reason => { // rejection
    console.error(reason.message);
});
```

### Assigning Rejection Handlers with `catch()`

Promises also have a `catch()` method that behaves the same as `then()` when only a rejection handler is passed. For example, the following `catch()` and `then()` calls are functionally equivalent:

```js
const promise = fetch('book.json');

promise.catch((reason) => {
	// rejection
	console.error(reason.message);
});
// is the same as:


promise.then(null, (reason) => {
	// rejection
	console.error(reason.message);
})
```

### Assigning Settlement Handlers with `finally()`
The callback passed to `finally()` (called a settlement handler) is called regardless of success or failure. Unlike the callbacks for `then()` and `catch()`, `finally()` callbacks do not receive any arguments because it isn’t clear whether the promise was fulfilled or rejected.

```js
const promise = fetch('book.json');

promise.finally(()=>{
// no way to know if fulfilled or rejected
console.log("Settled");
});

//is similar to:

const callback = () => {
	console.log("Settled");
};

promise.then(callback,callback);
```

### Assigning Handlers to Settled Promises

```js
const promise = fetch('book.json');

//original fulfillment handler
promise.then(response => {
	console.log(response.status);

	// now add another
	promise.then(response => {
		console.log(response.statusText);
	})
})
```

In this code, the fulfillment handler adds another fulfillment handler to the same promise. The promise is already fulfilled at this point, so the new fulfillment handler is added to the microtask queue and called when ready. Rejection and settlement handlers work the same way.

### Handlers and Microtasks

JavaScript executed in the regular flow of a program is executed as a task, which is to say that the JavaScript runtime has created a new execution context and executes the code completely, exiting when finished.
As an example, an `onclick` handler for a button in a web page is executed as a task. When the button is clicked, a new task is created and the `onclick` handler is executed. Once complete, the JavaScript runtime waits for the next interaction to execute more code. Promise handlers, however, are executed in a different way.

All promise handlers, whether fulfillment, rejection, or settlement, are executed as microtasks inside of the JavaScript engine. Microtasks are queued and then executed immediately after the currently running task has completed, before the JavaScript runtime becomes idle. Calling then(), catch(), or finally() tells a promise to queue the specified microtasks once the promise is settled.

```js
setTimeout(()=>{
	console.log("timer");

	queueMicrotask(() => {
		console.log("microtask in timer");
	});
},0);

queueMicrotask(() => {
  console.log("microtask");
});

// Output:
// 1. microtask
// 2. timer
// 3. microtask in timer

// Even though the timer is set for a delay of 0 milliseconds, the microtask executes first, followed by the timer, followed by the microtask inside of the timer.
```

The most important thing to remember about microtasks, including all promise handlers, is that they are executed as soon as possible once a task is complete. This minimizes the amount of time between a promise settling and the reaction to the settling, making promises suitable for situations where runtime performance is important.

### Creating New (Unsettled) Promises

New promises are created using the Promise constructor. This constructor accepts a single argument: a function called the executor, which contains the code to initialize the promise. The executor is passed two functions named resolve() and reject() as arguments. Do something within the callback, perhaps async, then call resolve if everything worked, otherwise call reject.

Like `throw` in plain old JavaScript, it's customary, but not required, to reject with an Error object. The benefit of Error objects is they capture a stack trace, making debugging tools more helpful.

#Note :  _A promise can only be resolved once, so if you call resolve() more than once inside of an executor, every call after the first is ignored._

##### Executor Errors
If an error is thrown inside an executor, then the promise’s rejection handler is called. For example:
```js
const promise = new Promise((resolve,reject)=>{
  throw new Error("Uh oh!");
});
// is equivalent to

// There is an implicit try-catch inside every executor so that the error is caught and then passed to the rejection handler.
const promise = new Promise((resolve,reject)=>{
	try {
		throw new Error("Uh oh!");
	} catch (ex) {
		reject(ex);
	}
});

promise.catch(reason=>{
	console.log(reason.message); // "Uh oh!"
});
```

### Creating Settled Promises
Using `Promise.resolve()` or  `Promise.reject()` method accepts a single argument and returns a promise in the fulfilled state or the rejected state.

```js
const promise = Promise.resolve(42);

promise.then(value=>{
  console.log(value); // 42
})

const promise2 = Promise.reject(42);
```

### Non-Promise Thenables
Both `Promise.resolve()` and `Promise.reject()` also accept non-promise thenables as arguments. When passed a non-promise thenable, these methods create a new promise that is called after the `then()` function.

A non-promise thenable is created when an object has a `then()` method that accepts a resolve and a reject argument, like this:
```js
const thenable = {
	then(resolve, reject) {
		resolve(42); // or reject(42)
	}
}
const promise = Promise.resolve(thenable);
promise.then((value) => {
  console.log(value) // 42
})
// or
promise.catch((value) => {
  console.log(value) // 42
})
```

In this example, Promise.resolve() calls thenable.then() so that a promise state can be determined.

###  Summary

A promise is a placeholder for a value that may be provided later as the result of some asynchronous operation. Instead of assigning an event handler or passing a callback into a function, you can use a promise to represent the result of an operation.

Promises have three states: pending, fulfilled, and rejected. A promise starts in a pending (unsettled) state and becomes fulfilled on a successful execution or rejected on a failure (fulfillment and rejection are settled states). In either case, handlers can be added to indicate when a promise is settled. The then() method allows you to assign a fulfillment and rejection handler; the catch() method allows you to assign only a rejection handler; the finally() method allows you to assign a settlement handler that is always called regardless of the outcome. All promise handlers are run as microtasks so they will not execute until the current script job is complete.

You can create new unsettled promises using the Promise constructor, which accepts an executor function as its only argument. The executor function is passed resolve() and reject() functions that you use to indicate the success or failure of the promise. The executor runs immediately upon creation of the promise, unlike handlers, which are run as microtasks. Any errors thrown in an executor are automatically caught and passed to reject().

While creating single promises is a useful and effective way to work with asynchronous operations in JavaScript, promises allow interesting composition patterns when chained together. In the next chapter, you’ll learn how promise handlers work to create promise chains and why that’s a valuable capability.


----
Reference:
- https://web.dev/promises/
- https://web.dev/async-functions/
